### ISSUE TITLE
【论文复现】torch.utils.data.DataLoader

### ISSUE THREAD

**Issue by liu824** [PaddlePaddle/Paddle#26701]
关于下面的语句该如何改为paddle版本的
torch.utils.data.DataLoader(dataseta, batch_size=batch_size, shuffle=True, num_workers=8, pin_memory=True)
我参考了DataLoader如下：
dl=layers.shuffle(fluid.io.DataLoader.from_generator(feed_list=dataseta,capacity=batch_size, use_multiprocess=True))
不知道这样使用是否正确，issue中也有用batch的，但是batch并没有设置线程数num_workers的地方

**Comment by DrRyanHuang** [comment]
我自己写了一个，供参考：
```python
class DataLoader():
    
    def __init__(self, 
                 dataset, 
                 transforms, # 感觉在 `fluid.io.xmap_readers` 中可以加速, 故而将 `transform` 提出来
                 sampler=None,
                 batch_size=2,
                 shuffle=True,
                 num_workers=1, # 这里不能改成 0
                 buffer_size=None,
                 drop_last=False):
        
        '''
        @Param
            transforms : `callable`
            dataset    : `callable` 同时返回一个 `generator`
        '''
        
        self.dataset = dataset
        self.transforms = transforms
        
        # `sampler` 的优先级高于 `shuffle`
        self.shuffle = shuffle if sampler is None else False 
        self.sampler = sampler
        
        # 以下交给 paddle 的参数即可
        self.batch_size = batch_size     
        self.num_workers = num_workers
        self.drop_last = drop_last
        self.buffer_size = cpu_count() if buffer_size is None else buffer_size  
        
        
        # 进行 `shuffle` 操作
        if self.shuffle:
            self.dataset._shuffle()
    
        self.xmap_reader = fluid.io.xmap_readers(self._mapper, 
                                                 self._reader, 
                                                 self.num_workers, 
                                                 self.buffer_size)
        self.batch_reader = fluid.io.batch(self.xmap_reader, 
                                           batch_size=self.batch_size, 
                                           drop_last=self.drop_last)
        
    # ---------------- magic func ----------------
    def __call__(self):
        return self.batch_reader()
    
    # ---------------- 自定义 func (不应外部调用) ----------------
    def _reader(self):
        if self.sampler is None:
            return self.dataset
        def sampler_reader():
            for idx in self.sampler:
                yield self.dataset[idx]
        return sampler_reader()    # 注意 `call` `sampler_reader` 才返回生成器
    
    def _mapper(self, x):
        temp = self.transforms(x[0])
        return temp, x[1]
```

**Comment by liu824** [comment]
谢谢大哥

**Comment by chenwhql** [comment]
https://www.paddlepaddle.org.cn/documentation/docs/en/api/io/DataLoader.html
其实这个已经有了，官网中文文档缺失，我们反馈一下，这里是英文的，可以先参考下

**Comment by liu824** [comment]
好的
